machine/machine_core/ssh_connection.py

```diff
@@ -417,16 +417,20 @@ def write(self, dest, content, append=False, owner=None, perm=None):
             append: If True, append to existing file instead of replacing it
             owner: If set, call chown on the file with the given owner string
             perm: Optional file permission as chmod shell string (e.g. "0600")
+
+        The directory of dest is created automatically.
         """
         assert dest
         assert self.ssh_address
 
-        cmd = "cat %s '%s'" % (append and '>>' or '>', dest)
+        dest_quoted = shlex.quote(dest)
+
+        self.execute(f"mkdir -p {shlex.quote(os.path.dirname(dest))}")
+        self.execute(f"cat {append and '>>' or '>'} {dest_quoted}", input=content)
         if owner:
-            cmd += " && chown '%s' '%s'" % (owner, dest)
+            self.execute(f"chown {owner} {dest_quoted}")
```

Comment:
Sorry to do this, but why not `['chown', owner, dest]`?

Reply:
shell is the enemy :shrug:

Reply:
done